### PR TITLE
[core][scheduler] simplify and improve scheduler

### PR DESCRIPTION
**Pull request by youkaichao**
fix the problem find in https://github.com/vllm-project/vllm/pull/6865#issuecomment-2254226713 .

avoid copy the whole queue in every step. use inplace update of the queue instead.

remove sorting policy, as the queues are always sorted by first-come-first-serve order.

### CODE REVIEW

**Comment by youkaichao** [comment]
Combining https://github.com/vllm-project/vllm/pull/6865 and this PR:

```shell
$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 1000

# 1k requests
Throughput: 694.62 requests/s, 22227.94 tokens/s

$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 10000

# 10k requests
Throughput: 735.86 requests/s, 23547.52 tokens/s

$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 100000

# 100k requests
Throughput: 656.60 requests/s, 21011.07 tokens/s

$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 500000

# 500k requests
Throughput: 609.16 requests/s, 19493.24 tokens/s
```

The throughput is almost constant w.r.t. the number of requests now.

**Comment by njhill** [comment]
@youkaichao could we keep the policy abstraction and just not use it in the default FCFS case? I know of research teams that are experimenting with different policies using this. Of course we can document that there may be nontrival cost to using a different policy.

**Comment by youkaichao** [comment]
> @youkaichao could we keep the policy abstraction and just not use it in the default FCFS case? I know of research teams that are experimenting with different policies using this. Of course we can document that there may be nontrival cost to using a different policy.

I don't want to support this. In my opinion, request prioritization should happen one layer above vLLM, e.g. in load balance layer and cross-instance scheduling layer.

vLLM is an inference engine, it's goal is to give results as quickly as possible. It does not make any sense to consider request priority here.

**Comment by njhill** [comment]
@youkaichao that's a valid opinion but I think we should collect others before making a decision on this.

I think this is an important PR/optimization but removal of the Policy abstraction is a separate question that should be addressed in a separate issue/PR imo.

**Comment by youkaichao** [comment]
Actually the policy is never feature-complete. It just applies to running queue and swapped queue. No policy is applied to the waiting queue.

If you want a full-fledged research framework for request prioritization, I don't think vLLM is a good palce.

For example, if a request joins, with highest priority, do you need to stop or preempt the running requests? When you swap out a request, and swap it in again, do you change it's priority? Is request priority a constant (determined when received), or can be dynamic and changing in every iteration?

With all these unclear questions, and the current naive implementation of policy in vLLM, I don't think it is worth maintaining, especially given that we need to aim for performance now.

**Comment by njhill** [comment]
@youkaichao vLLM does already serve as a foundation for many different research activities.

Re prioritization features, see https://github.com/vllm-project/vllm/issues/6077, https://github.com/vllm-project/vllm/pull/5958. @apatke @saurabhjha1 may want to comment further!

**Comment by youkaichao** [comment]
As discussed in #6077 , per-iteration sorting can be removed. And we will introduce priority interface later. So this PR is good to go now.

File: tests/core/block/e2e/test_correctness.py [hunk 1]
```diff
@@ -183,7 +183,7 @@ def test_v1_v2_greedy_equality_with_cow(baseline_llm_generator,
 
             # Allow only 2 sequences of ~128 tokens in worst case.
             # Note 16 = 128/block_size
-            "num_gpu_blocks_override": 2 * (16 + 1),
+            "num_gpu_blocks_override": 2 * (16 + 2),
```
Comment (WoosukKwon) [inline tests/core/block/e2e/test_correctness.py:186]: Just curious: What is this change for?

**Review by WoosukKwon — APPROVED** [review]
@youkaichao Thanks for this simplification!

File: tests/core/block/e2e/test_correctness.py (hunk 1 above)
Comment (youkaichao) [reply tests/core/block/e2e/test_correctness.py:186]: This is some corner case, that swapping order change in this PR makes the number of blocks not enough.

Comment (WoosukKwon) [reply tests/core/block/e2e/test_correctness.py:186]: @youkaichao Oh why the swapping order changes? I thought the change in this PR doesn't actually change the FCFS logic at all.

Comment (youkaichao) [reply tests/core/block/e2e/test_correctness.py:186]: The running queue is FCFS. The swapping queue is somewhat complicated, when you swap in / swap out again and again (which happens in this test case).

That being said, I don't think we have strict FCFS even before this PR, when we consider swap in / swap out again and again. I think we can only guarantee strict FCFS after we use priority queue in the future.

**Pull request merged**